### DISCUSSION TITLE
Paginate by time period (e.g. day) instead of a preterdemined size

### DISCUSSION THREAD

**Discussion by marckohlbrugge** [ddnexus/pagy#466] [Q&A]
Rather than say load 25 items at a time, I want to paginate my ActiveRecord collection by a time period in reverse chronological order.

So page 1 wil show today's created records. Page 2 will show yesterday's created records, and so forth.

However, if there's only a handful of records available I might want to group them by week/month instread. For example if there's a filter applied narrowing down the records a lot, paginating by day might not make sense and I should paginate by week or even month instead. 

I implemented some logic for this before, but I'd like to use it with Pagy.

```
# frozen_string_literal: true

class SmartGroupedByTimePeriod
  attr_reader :collection

  def initialize(collection, column: :created_at, minimum_count: 5)
    @collection = collection
    @minimum_count = minimum_count
    @column = column
  end

  def by_time_period
    case time_period
    when :day
      by_day
    when :week
      by_week
    when :month
      by_month
    end
  end

  def by_day
    @collection.group_by do |s|
      (s[@column] || s.created_at).beginning_of_day
    end
  end

  def by_week
    @collection.group_by do |s|
      (s[@column] || s.created_at).beginning_of_week
    end
  end

  def by_month
    @collection.group_by do |s|
      (s[@column] || s.created_at).beginning_of_month
    end
  end

  def time_period
    @time_period ||= if @collection.limit(nil).offset(nil).where(@column => (1.week.ago)..DateTime::Infinity.new).count >= @minimum_count * 7
                       :day
                     elsif @collection.limit(nil).offset(nil).where(@column => (1.month.ago)..DateTime::Infinity.new).count >= @minimum_count * 30
                       :week
                     else
                       :month
                     end
  end
end
```

As you can see this takes a collection, and uses Ruby array methods to group the records. This is not ideal because it has to load in _all_ the records into memory.

Any ideas how I can do this properly using Pagy?

I did see the calendar extra, which initially seemed relevant, but I think it's just for filtering on a specific time period whereas I'm talking about paginating all records by a time period.

**Comment by ddnexus** [comment]
> I did see the calendar extra, which initially seemed relevant, but I think it's just for filtering on a specific time period whereas I'm talking about paginating all records by a time period.

I don't get what you mean. Could you elaborate on that?
